### PR TITLE
Feature mssql

### LINKED ISSUE #1005: Collector and Analyzer to check MS SQL connection and version.
**Describe the rationale for the suggested feature.**

We have checks for other databases, we had a customer request similar functionality for MS SQL databases as well.

**Describe the feature**

- A collector will gather information from a MS SQL database, including connection status and version.
- An analyzer will make sure that if an external MS SQL database is attempting to be used, that there is a valid connection and a proper version being used.

```yaml
apiVersion: troubleshoot.sh/v1beta2
 kind: Preflight
 metadata:
   name: sample
 spec:
   collectors:
     - mssql:
         collectorName: mssql
         uri: sqlserver://username:password@hostname:1433/defaultdb
   analyzers:
     - mssql:
         checkName: Must be SQLServer 15.x or later
         collectorName: mssql
         outcomes:
           - fail:
               when: "connected == false"
               message: Cannot connect to SQLServer
           - fail:
               when: "version < 15.x"
               message: The SQLServer must be at least version 15
           - pass:
               message: The SQLServer connection checks out
```

**Describe alternatives you've considered**

**Additional context**

https://github.com/replicatedhq/troubleshoot/pull/990

### PR DESCRIPTION
**Pull request by chasehainey**
## Description, Motivation and Context

Please include a summary of the change or what problem it solves. Please also include relevant motivation and context.

This is a new collector and analyzer for MS SQL to be able to check for connections and ensure proper versions are being used.

## Checklist

- [x] New and existing tests pass locally with introduced changes.
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] The commit message(s) are informative and highlight any breaking changes
- [x] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by xavpaice** [comment]
you can run 'make schemas' and commit the result to fix the test that's failing, once you've verified that the changes are expected.

Is there a docs PR to go with this?  Would be good to link it here.

**Comment by chasehainey** [comment]
@xavpaice Still seeing the error: 
```
go build -ldflags " -s -w -X github.com/replicatedhq/troubleshoot/pkg/version.version=`git describe --tags --dirty` -X github.com/replicatedhq/troubleshoot/pkg/version.gitSHA=`git rev-parse HEAD` -X github.com/replicatedhq/troubleshoot/pkg/version.buildTime=`date -u +"%Y-%m-%dT%H:%M:%SZ"` " -o bin/schemagen github.com/replicatedhq/troubleshoot/cmd/schemagen
fatal: No names found, cannot describe anything.
./bin/schemagen --output-dir ./schemas
```
after running make schemas.

**Comment by chasehainey** [comment]
PR for troubleshoot.sh docs is [here](https://github.com/replicatedhq/troubleshoot.sh/pull/470).

File: examples/preflight/mssql.yaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  analyzers:
+    - mssql:
+        checkName: Must be SQLServer 15.x or later
+        collectorName: mssql
+        outcomes:
+          - fail:
+              when: "connected == false"
+              message: Cannot connect to SQLServer
+          - fail:
+              when: "version < 15.x"
```
Comment (banjoh) [inline examples/preflight/mssql.yaml:19]: Will `"version < 15.0.2000.x"` be evaluated correctly?

**Review by banjoh — DISMISSED** [review]

**Comment by banjoh** [comment]
Some tests for the analyser would be good as well

**Review by xavpaice — DISMISSED** [review]
I'd like to request tests (see Evans' comment), prior to merging.

File: examples/preflight/mssql.yaml (hunk 1 above)
Comment (chasehainey) [reply examples/preflight/mssql.yaml:19]: This has been handled and all unit tests are currently passing and evaluating the wildcards appropriately.

**Review by CpuID — DISMISSED** [review]
It looks like there's a fair few changes in this PR that aren't specifically related to the MSSQL collector, that are more improvements/tweaks to troubleshoot itself (and the CI processes).

Would it be possible to split those out into a separate PR, to make it easier to review MSSQL vs everything else independently? Assuming they are still relevant changes and it's not just a situation of needing a rebase or whatever... :)

**Comment by crdant** [comment]
> It looks like there's a fair few changes in this PR that aren't specifically related to the MSSQL collector, that are more improvements/tweaks to troubleshoot itself (and the CI processes).

I'm a little bit worried about this, @chasehainey @CpuID. There shouldn't be any. My concern would be that we could be introducing a regression rather than unintended improvements. All we did was rebase off of `main` and resolve merge confilicts.

**Comment by CpuID** [comment]
> > It looks like there's a fair few changes in this PR that aren't specifically related to the MSSQL collector, that are more improvements/tweaks to troubleshoot itself (and the CI processes).
> 
> 
> 
> I'm a little bit worried about this, @chasehainey @CpuID. There shouldn't be any. My concern would be that we could be introducing a regression rather than unintended improvements. All we did was rebase off of `main` and resolve merge confilicts.
> 
> 

Yea I agree it probably warrants investigation further in that case 👍

File: pkg/collect/mssql.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	"regexp"
+
+	_ "github.com/microsoft/go-mssqldb"
+	"github.com/pkg/errors"
+	troubleshootv1beta2 "github.com/replicatedhq/troubleshoot/pkg/apis/troubleshoot/v1beta2"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+<<<<<<< HEAD
```
Comment (banjoh) [inline pkg/collect/mssql.go]: @chasehainey you left some merge conflicts behind

**Review by banjoh — CHANGES_REQUESTED** [review]

File: pkg/collect/mssql.go (hunk 2 above)
Comment (chasehainey) [reply pkg/collect/mssql.go]: Thanks! I'm not sure why my editor didn't alert me that those were still there.

**Comment by banjoh** [comment]
You'll need to run `go mod tidy` as the failing checks suggest

**Comment by chasehainey** [comment]
Sorry! A regression made its way back in. I've fixed it and now it looks like our tests are passing.

**Review by banjoh — APPROVED** [review]

**Comment by banjoh** [comment]
I'll resolve your comments @CpuID & @xavpaice so as to proceed merging this change

**Review by CpuID — APPROVED** [review]
LGTM no objections and the tests pass

haven't tested it locally (don't have any MSSQL around easily to test it on)

**Comment by chris-sanders** [comment]
The documentation task is checked @chasehainey can you add a link to the PR where the docs changes exist so that these two can easily be referenced between themself?

**Comment by chasehainey** [comment]
@chris-sanders They've already been merged, but not added to the sidebar. https://github.com/replicatedhq/troubleshoot.sh/blob/564a1ef2d8237fc8536b06fc87668b1d3ee583c6/docs/source/analyze/mssql.md https://github.com/replicatedhq/troubleshoot.sh/blob/564a1ef2d8237fc8536b06fc87668b1d3ee583c6/docs/source/collect/mssql.md

**Pull request merged**